### PR TITLE
balancerd: initial metrics and liveness

### PR DESCRIPTION
**Pull request by mjibson**
Add a third listening port for internal observability. Hook up some first metrics.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Review by ParkMyCar — APPROVED** [review]
Looks good! Just one comment on that metrics edge case, but not going to block on that

File: src/balancerd/src/lib.rs [hunk 1]
```diff
@@ -244,10 +288,95 @@ impl BalancerService {
...
+
+impl ServerMetricsConfig {
+    fn register_into(registry: &MetricsRegistry) -> Self {
+        let connection_status = registry.register(metric!(
+            name: "mz_balancer_connections_status",
+            help: "Count of completed network connections, by status",
+            var_labels: ["source", "status"],
+        ));
+        let active_connections = registry.register(metric!(
+            name: "mz_balancer_connections_active",
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/balancerd/src/lib.rs:326]: My thought was that both this and the previous metric would share the same `mz_balancer_connection_` prefix. Thoughts?

File: src/balancerd/src/lib.rs [hunk 2]
```diff
@@ -371,10 +500,12 @@ impl mz_server_core::Server for PgwireBalancer {
     fn handle_connection(&self, conn: TcpStream) -> mz_server_core::ConnectionHandler {
         let tls = self.tls.clone();
         let resolver = Arc::clone(&self.resolver);
+        let metrics = self.metrics.clone();
         Box::pin(async move {
             // TODO: Try to merge this with pgwire/server.rs to avoid the duplication. May not be
             // worth it.
-            let _result: Result<(), anyhow::Error> = async move {
+            metrics.active_connections().inc();
+            let result: Result<(), anyhow::Error> = async move {
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/balancerd/src/lib.rs]: Neat! Done.

[the remark this answers is not in the document]
Comment (mjibson) [reply src/balancerd/src/lib.rs]: I taught the wrapper to inc itself which seemed handy.

File: src/balancerd/src/lib.rs [hunk 3]
```diff
@@ -244,10 +287,117 @@ impl BalancerService {
...
+#[derive(Clone, Debug)]
+struct ServerMetricsConfig {
+    connection_status: IntCounterVec,
+    active_connections: IntGaugeVec,
+}
+
+impl ServerMetricsConfig {
+    fn register_into(registry: &MetricsRegistry) -> Self {
+        let connection_status = registry.register(metric!(
+            name: "mz_balancer_connection_status",
```
Comment (pH14) [inline src/balancerd/src/lib.rs:340]: should `completed` be in the name here?

File: src/balancerd/src/lib.rs [hunk 4]
```diff
@@ -244,10 +287,117 @@ impl BalancerService {
...
+            help: "Count of completed network connections, by status",
+            var_labels: ["source", "status"],
+        ));
+        let active_connections = registry.register(metric!(
+            name: "mz_balancer_connection_active",
```
Comment (pH14) [inline src/balancerd/src/lib.rs:345]: would recommend using two counters: # of connections ever established, and # of connections completed. then the active count can be inferred by calculating them, while giving a bit more information overall.

Comment (pH14) [reply src/balancerd/src/lib.rs:345]: not sure what `source` is, but do we have any information on what the destination environment is for each connection? could be useful to surface that (at the risk of increased cardinality, but today I think that'd be fine)

**Review by pH14 — COMMENTED** [review]
Added a few belated comments! 🐌

**Pull request merged**